### ISSUE TITLE
Hh

### ISSUE THREAD

**Issue by micro521-coder** [micro521-coder/Nodejs#1]
IR PRACTICAL
Practical no 1
Aim: Document Indexing and Retrieval
• Implement an inverted index construction algorithm.
• Build a simple document retrieval system using the constructed index

# Define the documents
document1 = "The quick brown fox jumped over the lazy dog."
document2 = "The lazy dog slept in the sun."
# Step 1: Tokenize the documents
# Convert each document to lowercase and split it into words
tokens1 = document1.lower().split()
tokens2 = document2.lower().split()
# Combine the tokens into a list of unique terms
terms = list(set(tokens1 + tokens2))
# Step 2: Build the inverted index
# Create an empty dictionary to store the inverted index
inverted_index = {}
# For each term, find the documents that contain it
for term in terms:
    documents = []
    if term in tokens1:
        documents.append("Document 1")
    if term in tokens2:
        documents.append("Document 2")
    inverted_index[term] = documents
# Step 3: Print the inverted index
for term, documents in inverted_index.items():
    print(term, "->", ",".join(documents))

Practical no 2
Aim: Retrieval Models
• Implement the Boolean retrieval model and process queries.
• Implement the vector space model with TF-IDF weighting and cosine
similarity.

import pandas
from contextlib import redirect_stdout

terms = []
keys = []
vec_Dic = {}
dicti = {}
dummy_List = []

# Function to extract terms and documents
def filter(documents, rows, cols):
    for i in range(rows):
        for j in range(cols):
            if j == 0:
                keys.append(documents.loc[i].iat[j])
            else:
                dummy_List.append(documents.loc[i].iat[j])
                if documents.loc[i].iat[j] not in terms:
                    terms.append(documents.loc[i].iat[j])
        copy = dummy_List.copy()
        dicti.update({documents.loc[i].iat[0]: copy})
        dummy_List.clear()

# Function to convert documents to boolean vectors
def bool_Representation(dicti, rows, cols):
    terms.sort()
    for i in dicti:
        for j in terms:
            if j in dicti[i]:
                dummy_List.append(1)
            else:
                dummy_List.append(0)
        copy = dummy_List.copy()
        vec_Dic.update({i: copy})
        dummy_List.clear()

# Function to convert query to boolean vector
def query_Vector(query):
    qvect = []
    for i in terms:
        if i in query:
            qvect.append(1)
        else:
            qvect.append(0)
    return qvect

# Function to predict relevant document using Boolean matching
def prediction(q_Vect):
    dictionary = {}
    listi = []
    count = 0
    term_Len = len(terms)

    for i in vec_Dic:
        for t in range(term_Len):
            if q_Vect[t] == vec_Dic[i][t]:
                count += 1
        dictionary.update({i: count})
        count = 0

    for i in dictionary:
        listi.append(dictionary[i])

    listi = sorted(listi, reverse=True)
    ans = ''

    with open('output.txt', 'w') as f:
        with redirect_stdout(f):
            print("ranking of the documents")
            for count, i in enumerate(listi):
                key = check(dictionary, i)
                if count == 0:
                    ans = key
                print(key, "rank is", count + 1)
                dictionary.pop(key)
            print(ans, "is the most relevant document for the given query")

# Function to get document name by score
def check(dictionary, val):
    for key, value in dictionary.items():
        if val == value:
            return key

# Main function
def main():
    documents = pandas.read_csv(r'C:\Users\Admin\Desktop\SSINHA\documents - Sheet1.csv')
    rows = len(documents)
    cols = len(documents.columns)
    filter(documents, rows, cols)
    bool_Representation(dicti, rows, cols)
    print("Enter query")
    query = input()
    query = query.split(' ')
    q_Vect = query_Vector(query)
    prediction(q_Vect)

main()

Practical 3 
Aim Spelling Correction in IR Systems
• Develop a spelling correction module using edit distance algorithms.
• Integrate the spelling correction module into an information retrieval
system.

# Importing necessary libraries
import nltk
from nltk.metrics.distance import edit_distance
from nltk.corpus import words

# Downloading and importing the 'words' corpus
nltk.download('words')

# List of correct words from the 'words' corpus
correct_words = words.words()

# List of incorrect spellings that need to be corrected
incorrect_words = ['happpy', 'azmaing', 'intelliengt', 'natuer', 'ashy']

# Printing the incorrect words
print("Incorrect Words:", incorrect_words)
print("========= Result =========")

# Loop to find correct spellings based on edit distance
for word in incorrect_words:
    # Calculate the edit distance between the word and all correct words
    temp = [(edit_distance(word, w), w) for w in correct_words]
    
    # Print the closest correct word (sorted by minimum edit distance)
    corrected_word = sorted(temp, key=lambda val: val[0])[0][1]
    print(f"Incorrect word: {word} => Corrected word: {corrected_word}")

practical 4

PRECISION
from sklearn.metrics import precision_score

# Define actual values
act_pos = [1 for _ in range(100)]       # 100 actual positives
act_neg = [0 for _ in range(10000)]     # 10,000 actual negatives
y_true = act_pos + act_neg              # Combine both

# Define predicted values
pred_pos = [0 for _ in range(10)] + [1 for _ in range(90)]          # 90 predicted positives (10 false negatives)
pred_neg = [1 for _ in range(30)] + [0 for _ in range(9970)]        # 30 false positives
y_pred = pred_pos + pred_neg                                        # Combine both

# Calculate precision
precision = precision_score(y_true, y_pred, average='binary')
print('Precision: %.3f' % precision)

RECALL
from sklearn.metrics import recall_score

# define actual
act_pos = [1 for _ in range(100)]
act_neg = [0 for _ in range(10000)]
y_true = act_pos + act_neg

# define predictions
pred_pos = [0 for _ in range(10)] + [1 for _ in range(90)]
pred_neg = [0 for _ in range(10000)]
y_pred = pred_pos + pred_neg

# calculate recall
recall = recall_score(y_true, y_pred, average='binary')
print('Recall: %.3f' % recall)

F1-SCORE
from sklearn.metrics import f1_score

# define actual labels
act_pos = [1 for _ in range(100)]  # 100 positive instances
act_neg = [0 for _ in range(10000)]  # 10000 negative instances
y_true = act_pos + act_neg  # combine actual positive and negative labels

# define predictions
pred_pos = [0 for _ in range(5)] + [1 for _ in range(95)]  # 95 positive predictions, 5 negative
pred_neg = [1 for _ in range(55)] + [0 for _ in range(9945)]  # 55 positive predictions, 9945 negative predictions
y_pred = pred_pos + pred_neg  # combine predicted positive and negative labels

# calculate F1 score
score = f1_score(y_true, y_pred, average='binary')

# print the F1 score (F-measure)
print('F-Measure: %.3f' % score)

PRACTICAL 5 
Aim: Text Categorization • Implement a text classification algorithm (e.g., Naive Bayes or Support Vector Machines). • Train the classifier on a labelled dataset and evaluate its performance. 

import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

# Sample dataset
texts = [
    "I love this movie, it's amazing!",
    "What a great day, I feel so happy!",
    "This is the worst product I've ever bought.",
    "I'm so sad and disappointed.",
    "Absolutely fantastic experience, will buy again!",
    "Terrible customer service, very unhappy.",
    "I'm excited about this new opportunity!",
    "The food was awful and I got sick."
]

# Corresponding labels (1 for positive, 0 for negative)
labels = [1, 1, 0, 0, 1, 0, 1, 0]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    texts, labels, test_size=0.25, random_state=42
)

# Convert text data into feature vectors
vectorizer = CountVectorizer()
X_train_vectors = vectorizer.fit_transform(X_train)
X_test_vectors = vectorizer.transform(X_test)

# Train a Naive Bayes classifier
classifier = MultinomialNB()
classifier.fit(X_train_vectors, y_train)

# Make predictions
y_pred = classifier.predict(X_test_vectors)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy:.2f}")
print("Classification Report:")
print(report)

PRACTICAL 6
Aim: Clustering for Information Retrieval • Implement a clustering algorithm (e.g., K-means or hierarchical clustering). • Apply the clustering algorithm to a set of documents and evaluate the clustering results

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Load the dataset
dataset = pd.read_csv("D:\irpractice\Mall_Customers.csv")
print(dataset)

# Select relevant features (columns 3 and 4: Annual Income and Spending Score)
x = dataset.iloc[:, [3, 4]].values
print(x)

# Import KMeans
from sklearn.cluster import KMeans

# Find the optimal number of clusters using the Elbow Method
wcss_list = []  # WCSS = Within-Cluster Sum of Squares
for i in range(1, 11):
    kmeans = KMeans(n_clusters=i, init='k-means++', random_state=42)
    kmeans.fit(x)
    wcss_list.append(kmeans.inertia_)

# Plotting the Elbow Graph
plt.plot(range(1, 11), wcss_list)
plt.title("The Elbow Method Graph")
plt.xlabel("Number of clusters (k)")
plt.ylabel("WCSS (Within-Cluster Sum of Squares)")
plt.show()
PRACTICAL 7
Aim: Web Crawling and Indexing • Develop a web crawler to fetch and index web pages. • Handle challenges such as robots.txt, dynamic content, and crawling delays.

import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin

# Simple Web Crawler
class SimpleWebCrawler:
    def __init__(self, base_url, max_pages=10):
        self.base_url = base_url
        self.max_pages = max_pages
        self.visited = set()

    def crawl(self, url, depth=0):
        if url in self.visited or depth >= self.max_pages:
            return
        try:
            response = requests.get(url)
            if response.status_code != 200:
                return
            print(f"Crawling: {url}")
            self.visited.add(url)
            soup = BeautifulSoup(response.text, 'html.parser')
            for link in soup.find_all('a', href=True):
                next_url = urljoin(url, link['href'])
                if next_url.startswith(self.base_url):
                    self.crawl(next_url, depth + 1)
        except Exception as e:
            print(f"Failed to crawl {url}: {e}")

# Usage example
if __name__ == "__main__":
    start_url = "https://www.tpointtech.com/"
    crawler = SimpleWebCrawler(base_url=start_url, max_pages=5)
    crawler.crawl(start_url)

PRACTICAL 8
Aim: - Link Analysis and PageRank • Implement the PageRank algorithm to rank web pages based on link analysis. • Apply the PageRank algorithm to a small web graph and analyze the results. 

import numpy as np

def pagerank(graph, alpha=0.85, tol=1.0e-6, max_iter=100):
    """Computes PageRank scores for a directed graph."""
    n = len(graph)
    if n == 0:
        return {}

    # Initialize ranks
    ranks = np.ones(n) / n

    # Create transition matrix M
    M = np.zeros((n, n))

    # Construct the transition matrix
    for i, node in enumerate(graph):
        links = graph[node]
        if links:
            M[i, [list(graph.keys()).index(dest) for dest in links]] = 1 / len(links)
        else:
            M[i, :] = 1 / n  # Handling dangling nodes

    # Power iteration method
    for _ in range(max_iter):
        new_ranks = alpha * np.dot(M.T, ranks) + (1 - alpha) / n * np.ones(n)
        if np.linalg.norm(new_ranks - ranks, ord=1) < tol:
            break
        ranks = new_ranks

    return {node: ranks[i] for i, node in enumerate(graph)}

if __name__ == "__main__":
    # Example web graph
    web_graph = {
        "A": ["B", "C"],
        "B": ["C", "D"],
        "C": ["A"],
        "D": ["C"]
    }

    # Compute PageRank
    ranks = pagerank(web_graph)

    # Print PageRank Scores
    print("PageRank Scores:")
    for page, rank in ranks.items():
        print(f"{page}: {rank:.4f}")

PRACTICAL 9
Aim: - Advanced Topics in Information Retrieval • Implement a text summarization algorithm (e.g., extractive or abstractive). • Build a question-answering system using techniques such as information extraction 

CODE 1
from transformers import pipeline

def abstractive_summary(text, min_length=30, max_length=130):
    """
    Generate an abstractive summary of the given text using Hugging Face's transformers pipeline.

    Parameters:
    - text (str): The input text to summarize.
    - min_length (int): Minimum length of the summary.
    - max_length (int): Maximum length of the summary.

    Returns:
    - str: The abstractive summary.
    """
    summarizer = pipeline("summarization")
    summary = summarizer(text, min_length=min_length, max_length=max_length)
    return summary[0]['summary_text']

# Example usage:
text = """
Artificial Intelligence (AI) is rapidly transforming various industries including healthcare, 
finance, education, and transportation. It enables machines to learn from data, adapt to new 
inputs, and perform tasks that usually require human intelligence. As AI continues to evolve, 
it promises to improve efficiency, accuracy, and decision-making across many sectors. 
However, it also raises concerns around job displacement, data privacy, and ethical use.
"""
print("Abstractive Summary:")
print(abstractive_summary(text))
CODE 2
from gensim.summarization import summarize

def extractive_summary(text, ratio=0.2):
    """
    Generate an extractive summary of the given text using Gensim's summarize function.

    Parameters:
    - text (str): The input text to summarize.
    - ratio (float): The ratio of sentences to include in the summary (default is 0.2).

    Returns:
    - str: The extractive summary.
    """
    try:
        summary = summarize(text, ratio=ratio)
        return summary
    except ValueError:
        return "Input text is too short to summarize."

# Example usage:
text = """
Artificial intelligence (AI) is intelligence demonstrated by machines, in contrast to the natural
intelligence displayed by humans. Leading AI textbooks define the
field as the study of "intelligent agents": any device that perceives its environment and takes
actions to maximize its chance of success at some goal. Colloquially, the term "artificial
intelligence" is applied when a machine mimics "cognitive" functions that humans associate
with the human mind, such as "learning" and "problem-solving". As machines become
increasingly capable, tasks considered to require "intelligence" are often removed from the
definition of AI. A quip in Tesler's Theorem says "AI is whatever hasn't been done yet."
For instance, optical character recognition is frequently excluded from things considered to
be AI, having become a routine technology.
Modern machine capabilities generally classified as AI include successfully understanding
human speech, competing at the highest level in strategic game systems (such as chess and
Go), self-driving cars, and more.
"""

print("Extractive Summary:")
print(extractive_summary(text))
